### DISCUSSION TITLE
Ask for confirmation when detach a script

### DISCUSSION THREAD

**Discussion by AgatheMametz** [godotengine/godot-proposals#5073] [Editor]
![Godot_v3 4 4_LMJ2nzKwXo](https://user-images.githubusercontent.com/45824145/182906272-d3301c24-5ea2-4118-8a2d-998a8d460dad.gif)

Hello,
I find it odd that you can detach a script from it's node in just one click without confirmation. If your script did not has the same name as its node, when you reclick on the script button you can't find it back easily.

My suggestion would be to add a modal to ask for confirmation when you try to detach a script.

It's my first proposition here so I hope I searched enough to be sure it was not proposed already.

**Comment by mieldepoche** [comment]
I'm not convinced this is needed since a modal is just more clicks to achieve something you can easily undo (Ctrl+Z).

> you can't find it back easily

You can search for it in the FileSystem dock and drag-and-drop it back on your node if this is too far away in the history.

**Comment by Calinou** [comment]
Detaching a script doesn't result in loss of data (unlike removing the script file from disk), so it shouldn't have a confirmation dialog.

Since 10 years ago or so, there's been a move to remove redundant confirmation dialogs from operating systems and applications. For example, this is why Windows 8 and later will no longer display a confirmation dialog when moving a file to trash. (You'll still see a dialog when emptying the trash or when using <kbd>Shift + Delete</kbd> to remove files permanently, as both of these operations cause loss of data.)

**Reply by wareya** [reply]
Uh, it *does* result in loss of data, if you have exported variables or you're using a built-in script with no other references. It's sometimes possible to recover from this using the undo/redo system, but the undo/redo system is very buggy w/r/t what happens to the stack when switching between scenes etc (sometimes you completely lose everything that was in the undo buffer with no way to recover it), and is currently slightly buggy for exported variable recovery when undoing the detachment of a script (detach -> undo doesn't bring exported variables back, but detach -> undo -> undo -> redo does). So the undo/redo system should not be relied for this.

**Reply by Calinou** [reply]
> but the undo/redo system is very buggy w/r/t what happens to the stack when switching between scenes etc (sometimes you completely lose everything that was in the undo buffer with no way to recover it),

This should be fixed by https://github.com/godotengine/godot/pull/59564.
